modules/javafx.graphics/src/main/native-prism-d3d/D3DMeshView.cc

```diff
@@ -145,92 +137,110 @@ void D3DMeshView::render() {
     }
 
     computeNumLights();
-    // We only support up to 3 point lights at the present
-    for (int i = 0; i < MAX_NUM_LIGHTS; i++) {
-        status = SUCCEEDED(device->SetVertexShaderConstantF(VSR_LIGHTS + i*2, lights[i].position, 1));
-    }
 
-    float lightsNormDirection[MAX_NUM_LIGHTS * 4]; // 3 lights x (3 coords + 1 padding)
-    for (int i = 0, d = 0; i < MAX_NUM_LIGHTS; i++) {
-        if (lights[i].isPointLight()) {
-            lightsNormDirection[d++] = 0;
-            lightsNormDirection[d++] = 0;
-            lightsNormDirection[d++] = 1;
-            lightsNormDirection[d++] = 0;
+    // Prepare lights data
+    float lightsPosition[MAX_NUM_LIGHTS * 4];      // 3 coords + 1 padding
+    float lightsNormDirection[MAX_NUM_LIGHTS * 4]; // 3 coords + 1 padding
+    float lightsColor[MAX_NUM_LIGHTS * 4];         // 3 color + 1 padding
+    float lightsAttenuation[MAX_NUM_LIGHTS * 4];   // 3 attenuation factors + 1 isAttenuated
+    float lightsRange[MAX_NUM_LIGHTS * 4];         // 1 maxRange + 3 padding
+    float spotLightsFactors[MAX_NUM_LIGHTS * 4];   // 2 angles + 1 falloff + 1 padding
+    for (int i = 0, d = 0, p = 0, c = 0, a = 0, r = 0, s = 0; i < MAX_NUM_LIGHTS; i++) {
+        D3DLight light = lights[i];
```

Comment:
You're invoking the auto-generated copy constructor of `D3DLight` here, where the original code didn't do that. Just making sure that that's what you intended.

Reply:
I will change to `D3DLight& light = lights[i];`.